src/core/qgsdatasourceuri.cpp

```diff
@@ -864,7 +864,7 @@ void QgsDataSourceUri::setParam( const QString &key, const QString &value )
   else
   {
     // may be multiple
-    mParams.insert( key, value );
+    mParams.insert( key, QUrl::toPercentEncoding( value ) );
```

Comment:
I dind't dig deep into this issue but I believe that if possible you should leave the stored value unencoded and make sure it is correctly encoded before it is used inside an URL or marshalled to a permanent storage.

Generally speaking it is a better practice to store raw values and encode them (when necessary) right before usage.

Reply:
I have moved the encoding into QgsDataSourceUri::encodedUri() now and it appears to still fix the encoding issue while being more backwards compatible.
Also it seems to adhere to the way QUrlQuery is supposed to be used according to the documentation.

> All of the setter methods and the query methods like [hasQueryItem](https://doc.qt.io/qt-6/qurlquery.html#hasQueryItem)() in QUrlQuery take encoded forms only.

> Application code should strive to always ensure proper encoding and not rely on TolerantMode parsing fixing the strings. Notably, all user input must be first percent-encoded using [QUrl::toPercentEncoding](https://doc.qt.io/qt-6/qurl.html#toPercentEncoding)() or similar functions before being passed to the functions in this class.

(https://doc.qt.io/qt-6/qurlquery.html#encoding)

I will take a look at the tests now.